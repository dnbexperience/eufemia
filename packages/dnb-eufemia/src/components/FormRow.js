/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index form-row to autogenerate all the components and patterns
 * Used by "prepareFormRows"
 */

import FormRow from './form-row/FormRow'
export * from './form-row/FormRow'
export default FormRow
