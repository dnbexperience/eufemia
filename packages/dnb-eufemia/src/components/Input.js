/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index input to autogenerate all the components and patterns
 * Used by "prepareInputs"
 */

import Input from './input/Input'
export * from './input/Input'
export default Input
