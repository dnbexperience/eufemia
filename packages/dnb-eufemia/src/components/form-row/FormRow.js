/**
 * Web FormRow Component
 *
 * This is a legacy component.
 * For referencing while developing new features, please use a Functional component.
 */

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {
  extendGracefully,
  extendPropsWithContextInClassComponent,
  isTrue,
  makeUniqueId,
  validateDOMAttributes,
} from '../../shared/component-helper'
import AlignmentHelper from '../../shared/AlignmentHelper'

import Context from '../../shared/Context'
import FormLabel from '../form-label/FormLabel'
import Section from '../section/Section'
import {
  spacingPropTypes,
  createSpacingClasses,
} from '../space/SpacingHelper'

export const formRowPropTypes = {
  id: PropTypes.string,
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.node,
  ]),
  label_direction: PropTypes.oneOf(['vertical', 'horizontal']),
  label_sr_only: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  label_id: PropTypes.string,
  label_class: PropTypes.string,
  no_label: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  no_fieldset: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  locale: PropTypes.string,
  wrap: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  direction: PropTypes.oneOf(['vertical', 'horizontal']),
  vertical: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  centered: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  section_style: PropTypes.string,
  section_spacing: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  globalStatus: PropTypes.shape({
    id: PropTypes.string,
    message: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  }),
  responsive: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  disabled: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  skeleton: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

  skipContentWrapperIfNested: PropTypes.bool,

  ...spacingPropTypes,

  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.node,
  ]),
}

export const formRowDefaultProps = {
  id: null,
  label: null,
  label_direction: null,
  label_sr_only: null,
  label_id: null,
  label_class: null,
  no_label: false,
  no_fieldset: null,
  locale: null,
  wrap: null,
  direction: null,
  vertical: null,
  centered: null,
  section_style: null,
  section_spacing: null,
  globalStatus: null,
  responsive: null,
  disabled: null,
  skeleton: null,

  skipContentWrapperIfNested: false,
  className: null,
  children: null,
}

export default class FormRow extends React.PureComponent {
  static contextType = Context

  static propTypes = {
    ...formRowPropTypes,
  }

  static defaultProps = {
    ...formRowDefaultProps,
  }

  static getContent(props) {
    let label = null
    let children =
      typeof props.children === 'function'
        ? props.children(props)
        : props.children

    if (Array.isArray(props.children)) {
      children = children.reduce((pV, cV) => {
        if (cV && cV.type && cV.type.name === 'FormLabel') {
          label = cV.props.children
        } else {
          pV.push(cV)
        }
        return pV
      }, [])
    }

    return { label, children }
  }

  constructor(props) {
    super(props)
    this._id = props.id || makeUniqueId() // cause we need an id anyway

    // We may consider to use this later to check if we are inside FormSet
    // this.isInsideFormSet =
    //   context.FormRow && context.FormRow.isInsideFormSet
  }

  render() {
    // use only the props from context, who are available here anyway
    const props = extendPropsWithContextInClassComponent(
      this.props,
      FormRow.defaultProps,
      this.context?.FormRow, // nested FormRow
      this.context?.formElement // nested formElement
    )

    const {
      label_direction,
      label_sr_only,
      label_id,
      label_class,
      no_fieldset,
      no_label,
      locale,
      direction,
      vertical,
      centered,
      section_style,
      section_spacing,
      globalStatus,
      responsive,
      disabled,
      skeleton,
      wrap,
      className,
      skipContentWrapperIfNested,

      id: _id, // eslint-disable-line
      label: _label, // eslint-disable-line

      ...attributes
    } = props

    let { label } = props

    const isNested =
      this.context?.FormRow?.itsMeAgain ||
      this.context?.formElement?.itsMeAgain

    // in case we have a label already, we split this out and use this one instead
    const { label: nestedLabel, children } = FormRow.getContent(this.props)
    if (!label && nestedLabel) {
      label = nestedLabel
    }
    const hasLabel =
      (typeof label === 'string' && label.length > 0) || label
        ? true
        : false

    const id = this._id
    const params = {
      className: classnames(
        'dnb-form-row',
        (isTrue(vertical) || direction) &&
          `dnb-form-row--${isTrue(vertical) ? 'vertical' : direction}`,
        (isTrue(vertical) || label_direction) &&
          `dnb-form-row--${
            isTrue(vertical) ? 'vertical' : label_direction
          }-label`, // <-- has label
        centered && 'dnb-form-row--centered',
        isNested && 'dnb-form-row--nested',
        createSpacingClasses(props),
        className
      ),
      ...attributes,
    }

    // also used for code markup simulation
    validateDOMAttributes(this.props, params)

    const formElement = {
      useId: () => {
        if (this.isIsUsed) {
          // make a new ID, as we used one
          return makeUniqueId() // cause we need an id anyway
        }
        this.isIsUsed = true
        return id
      },
      itsMeAgain: true,
      hasLabel,
      globalStatus,
      direction,
      vertical,
      label_direction: isTrue(vertical) ? 'vertical' : label_direction,
      responsive,
      disabled,
      skeleton,
    }

    const providerContext = extendGracefully(this.context, {
      locale: locale ? locale : this.context.locale,
      formElement,
      FormRow: formElement,
    })

    const useFieldset = !isTrue(no_fieldset) && hasLabel

    return (
      <Context.Provider value={providerContext}>
        <Fieldset useFieldset={useFieldset}>
          <Section
            style_type={section_style || 'transparent'}
            spacing={section_spacing}
            element="div"
            {...params}
          >
            <AlignmentHelper />

            {label && (
              <FormLabel
                className={classnames('dnb-form-row__label', label_class)}
                id={label_id ? label_id : id + '-label'}
                for_id={useFieldset ? null : id} // we don't use for_id, because we don't have a single element to target to
                text={label}
                element={useFieldset ? 'legend' : 'label'}
                label_direction={label_direction}
                sr_only={label_sr_only}
                disabled={disabled}
                skeleton={skeleton}
              />
            )}

            {isTrue(no_label) && (
              <span
                className="dnb-form-label dnb-form-row__label-dummy"
                aria-hidden
              />
            )}

            {isNested && skipContentWrapperIfNested ? (
              children
            ) : (
              <div
                className={classnames(
                  'dnb-form-row__content',
                  isTrue(wrap) && 'dnb-form-row__content--wrap',
                  label && !isTrue(vertical) && direction !== 'vertical',
                  responsive && 'dnb-responsive-component'
                )}
              >
                {children}
              </div>
            )}
          </Section>
        </Fieldset>
      </Context.Provider>
    )
  }
}

const Fieldset = ({
  useFieldset,
  children,
  className = null,
  ...props
}) => {
  if (useFieldset) {
    return (
      <fieldset
        className={classnames('dnb-form-row__fieldset', className)}
        {...props}
      >
        {children}
      </fieldset>
    )
  }
  return (
    <div
      className={classnames('dnb-form-row__fieldset', className)}
      {...props}
    >
      {children}
    </div>
  )
}

// docs (or use ptd): https://github.com/facebook/prop-types#usage
Fieldset.propTypes = {
  children: PropTypes.node,
  useFieldset: PropTypes.bool,
  className: PropTypes.string,
}
Fieldset.defaultProps = {
  children: null,
  useFieldset: false,
  className: null,
}

FormRow._supportsSpacingProps = true
