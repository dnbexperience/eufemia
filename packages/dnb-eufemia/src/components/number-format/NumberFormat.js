/**
 * Web NumberFormat Component
 *
 * This is a legacy component.
 * For refferencing while developing new features, please use a Functional component.
 */

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import Context from '../../shared/Context'
import {
  warn,
  isTrue,
  makeUniqueId,
  validateDOMAttributes,
  convertJsxToString,
  extendPropsWithContextInClassComponent,
  extendDeep,
  detectOutsideClick,
} from '../../shared/component-helper'
import { hasSelectedText, IS_IOS } from '../../shared/helpers'
import {
  spacingPropTypes,
  createSpacingClasses,
} from '../space/SpacingHelper'
import {
  skeletonDOMAttributes,
  createSkeletonClass,
} from '../skeleton/SkeletonHelper'
import Tooltip, { injectTooltipSemantic } from '../tooltip/Tooltip'
import {
  format,
  showSelectionNotice,
  runIOSSelectionFix,
} from './NumberUtils'

export default class NumberFormat extends React.PureComponent {
  static contextType = Context

  static propTypes = {
    id: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    locale: PropTypes.string,
    prefix: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    suffix: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

    // currency
    currency: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    currency_display: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.oneOf(['code', 'name', 'symbol', 'narrowSymbol', '']),
    ]),
    currency_position: PropTypes.oneOf(['auto', 'before', 'after']),

    // shortens any number or currency including an abbreviation
    compact: PropTypes.oneOfType([
      PropTypes.oneOf(['short', 'long']),
      PropTypes.bool,
    ]),

    // bank account number
    ban: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // national identification number
    nin: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // phone number
    phone: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // organization number
    org: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // percentage
    percent: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // can be tel or sms
    link: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    options: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
    decimals: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    selectall: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    always_selectall: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]),
    copy_selection: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]),
    clean_copy_value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]),
    omit_rounding: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    clean: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    srLabel: PropTypes.node,
    element: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    tooltip: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.func,
      PropTypes.node,
    ]),
    skeleton: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    ...spacingPropTypes,

    class: PropTypes.string,
    className: PropTypes.string,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  }
  static defaultProps = {
    id: null,
    value: null,
    locale: null,
    prefix: null,
    suffix: null,
    currency: null,
    currency_display: null, // code, name, symbol
    currency_position: null, // null, before, after
    compact: null,
    ban: null,
    nin: null,
    phone: null,
    org: null,
    percent: null,
    link: null,
    options: null,
    decimals: null,
    selectall: true,
    always_selectall: false,
    copy_selection: true,
    clean_copy_value: false,
    omit_rounding: null,
    clean: null,
    srLabel: null,
    element: 'span', // span or abbr
    tooltip: null,
    skeleton: null,
    class: null,

    className: null,
    children: null,
  }

  constructor(props) {
    super(props)
    this._ref = React.createRef()
    this._selectionRef = React.createRef()

    this._id = props.tooltip ? props.id || makeUniqueId() : undefined
    this.state = { selected: false, omitCurrencySign: false, hover: false }
  }

  componentDidMount() {
    clearTimeout(this._selectAllTimeout)

    // NB: This hack may be removed in future iOS versions
    // in order that iOS v13 can select something on the first try, we run this add range trick
    if (IS_IOS && !hasiOSFix) {
      hasiOSFix = true
      runIOSSelectionFix()
    }
  }

  shortcutHandler = () => {
    showSelectionNotice({
      value: this.cleanedValue,
      label: this.context.getTranslation(this.props)?.NumberFormat
        ?.clipboard_copy,
    }).run(this._ref.current)
  }

  onBlurHandler = () => {
    this.setState({ selected: false })
  }

  onContextMenuHandler = () => {
    if (!hasSelectedText()) {
      clearTimeout(this._selectAllTimeout)
      this._selectAllTimeout = setTimeout(() => {
        this.setFocus()
      }, 1)
    }
  }

  onClickHandler = () => {
    if (
      (isTrue(this.props.selectall) ||
        isTrue(this.props.always_selectall)) &&
      !hasSelectedText()
    ) {
      this.setFocus()
    }
  }

  componentWillUnmount() {
    this.outsideClick?.remove()
  }

  setFocus() {
    this.setState({ selected: true }, () => {
      this._selectionRef.current?.focus()
      this.selectAll()

      if (!isTrue(this.props.copy_selection)) {
        this.outsideClick = detectOutsideClick(
          this._ref.current,
          this.onBlurHandler
        )
      }
    })
  }

  selectAll() {
    try {
      const elem = this._selectionRef.current || this._ref.current
      if (elem) {
        const selection = window.getSelection()
        const range = document.createRange()
        range.selectNodeContents(elem)
        selection.removeAllRanges()
        selection.addRange(range)
      }
    } catch (e) {
      warn(e)
    }
  }

  runFix(comp, className) {
    if (typeof comp === 'function') {
      comp = comp()
    }
    if (React.isValidElement(comp)) {
      return React.cloneElement(comp, {
        className: classnames(comp.props.className, className),
      })
    }
    return <span className={className}>{comp}</span>
  }

  onMouseEnter = () => {
    this.setState({ hover: true })
  }

  onMouseLeave = () => {
    this.setState({ hover: false })
  }

  render() {
    // consume the global context
    const props = extendPropsWithContextInClassComponent(
      this.props,
      NumberFormat.defaultProps,
      this.context.getTranslation(this.props).NumberFormat,
      this.context.NumberFormat
    )

    const {
      id, // eslint-disable-line
      value: _value,
      prefix,
      suffix,
      children,
      currency,
      currency_display,
      currency_position,
      compact,
      ban,
      nin,
      phone,
      org,
      percent,
      link: _link,
      tooltip,
      skeleton,
      options,
      locale,
      decimals,
      omit_rounding,
      clean,
      selectall,
      copy_selection,
      clean_copy_value,
      srLabel,
      element,
      className,
      ..._rest
    } = props
    let rest = _rest

    let link = _link
    let value = _value

    if (value === null && children !== null) {
      value = children
    }

    const formatOptions = {
      locale,
      currency,
      currency_display,
      currency_position,
      omit_currency_sign: this.state.omitCurrencySign,
      compact,
      ban,
      nin,
      phone,
      org,
      percent,
      decimals,
      omit_rounding: isTrue(omit_rounding),
      options,
      clean: isTrue(clean),
      clean_copy_value: isTrue(clean_copy_value),
      returnAria: true,
    }

    // use only the props from context, who are available here anyway
    const useContext = extendDeep(
      { locale: null, currency: null },
      this.context
    )

    if (useContext) {
      if (useContext.locale && !locale) {
        formatOptions.locale = useContext.locale
      }

      // only replace if the prop is "true" and not actually a currency
      if (useContext.currency && isTrue(currency)) {
        formatOptions.options = formatOptions.options
          ? { ...formatOptions.options }
          : {}
        formatOptions.options.currency = useContext.currency
      }
    }

    let {
      cleanedValue,
      number: display,
      aria,
      locale: lang,
    } = format(value, formatOptions)
    this.cleanedValue = cleanedValue

    if (prefix) {
      display = (
        <>
          {this.runFix(prefix, 'dnb-number-format__prefix')} {display}
        </>
      )
      aria = String(
        `${convertJsxToString(
          this.runFix(prefix, 'dnb-number-format__prefix')
        )} ${aria}`
      )
    }
    if (suffix) {
      display = (
        <>
          {display} {this.runFix(suffix, 'dnb-number-format__suffix')}
        </>
      )
      aria = `${aria} ${convertJsxToString(
        this.runFix(suffix, 'dnb-number-format__suffix')
      )}`
    }

    if (tooltip) {
      rest = injectTooltipSemantic(rest)
    }

    const attributes = {
      lang,
      ref: this._ref,
      className: classnames(
        'dnb-number-format',
        className,
        (isTrue(currency) || typeof currency === 'string') &&
          'dnb-number-format--currency',
        isTrue(selectall) && 'dnb-number-format--selectall',
        this.state.selected && 'dnb-number-format--selected',
        link && 'dnb-anchor',
        createSpacingClasses(this.props)
      ),

      // Makes it possible for NVDA to read on mouse over
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave,

      ...rest,
    }

    /**
     * Works in VoiceOver and NVDA
     * Makes the span with it's roles etc. appear as text.
     * Special useful if a number is in side e.g. a paragraph alongside with numbers
     */
    attributes['role'] = 'text' // role="text"

    const displayParams = {}
    if (isTrue(selectall) || isTrue(copy_selection)) {
      displayParams.onClick = this.onClickHandler
      displayParams.onContextMenu = this.onContextMenuHandler
    }

    validateDOMAttributes(this.props, attributes)
    skeletonDOMAttributes(attributes, skeleton, this.context)

    if (link) {
      if (isTrue(link)) {
        link = 'tel'
      }
      return (
        <a href={`${link}:${display}`} {...attributes}>
          {display}
        </a>
      )
    }

    const Element = element

    return (
      <Element {...attributes}>
        <span
          className={classnames(
            'dnb-number-format__visible',
            createSkeletonClass('font', skeleton, this.context)
          )}
          // Makes it possible for NVDA to read on mouse over
          aria-hidden={!this.state.hover}
          {...displayParams}
        >
          {display}
        </span>

        {/* Used for VoiceOver and NVDA when navigating with arrow keys */}
        <span
          className="dnb-sr-only"
          // Use "data-text" so Chrome does not copy the HTML as content, when pasting it in Outlook etc.
          data-text={
            srLabel ? `${convertJsxToString(srLabel)}${' '}${aria}` : aria
          }
        />

        {isTrue(copy_selection) && (
          <span
            className="dnb-number-format__selection dnb-no-focus"
            ref={this._selectionRef}
            tabIndex={-1}
            onBlur={this.onBlurHandler}
            onCopy={this.shortcutHandler}
            aria-hidden
          >
            {this.state.selected && cleanedValue}
          </span>
        )}

        {tooltip && (
          <Tooltip
            id={this._id + '-tooltip'}
            targetElement={this._ref}
            tooltip={tooltip}
          />
        )}
      </Element>
    )
  }
}

let hasiOSFix = false

NumberFormat._supportsSpacingProps = true
