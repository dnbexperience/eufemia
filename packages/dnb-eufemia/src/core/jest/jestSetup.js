/**
 * Jest Setup
 *
 */

import { axe, toHaveNoViolations } from 'jest-axe'
import fakeProps from 'react-fake-props'
import { shallow, mount, render } from './enzyme'
import ReactDOMServer from 'react-dom/server'
import fs from 'fs-extra'
import onceImporter from 'node-sass-once-importer'
import path from 'path'
import sass from 'node-sass'
import { toBeType } from 'jest-tobetype'
import toJson from 'enzyme-to-json'

export {
  fakeProps, // we have also our own replacement function called "fakeAllProps"
  shallow,
  mount,
  render,
  toJson,
  axe,
  toHaveNoViolations
}

global.IS_TEST = true
if (typeof window !== 'undefined') {
  window.IS_TEST = true
}

expect.extend({ toBeType })
expect.extend(toHaveNoViolations)

export const loadScss = (file, options = {}) => {
  try {
    const sassResult = sass.renderSync({
      file,
      includePaths: [path.resolve(__dirname, '../../style/core/')],
      importer: [onceImporter()],
      ...options
    })
    return String(sassResult.css)
  } catch (e) {
    console.error('loadScss error:', e)
    return e
  }
}

export const mockGetSelection = () => {
  let memory
  Object.defineProperty(window.navigator, 'clipboard', {
    configurable: true,
    value: {
      writeText: jest.fn().mockImplementation((v) => {
        memory = v
        return Promise.resolve(v)
      }),
      readText: jest.fn().mockImplementation(() => Promise.resolve(memory))
    }
  })

  const mockRange = new (class Range {
    constructor() {
      this.startContainer = {
        parentNode: document.createElement('div')
      }
    }
    getElement() {
      return this.node
    }
    insertNode(elem) {
      this.node = document.createElement('div')
      this.node.appendChild(elem)
      return this
    }
    cloneRange() {
      return this
    }
  })()

  let ranges = [mockRange]

  const mockValue = '1234.56'
  let value = mockValue
  let rangeCount = 9

  class RangeObj {
    rangeCount = rangeCount
    toString = () => value
    addRange(range = mockRange) {
      value = mockValue
      ranges.push(range)
      rangeCount = ranges.length
    }
    getRangeAt(index) {
      return ranges[index]
    }
    removeAllRanges() {
      value = ''
      ranges = []
      rangeCount = ranges.length
    }
  }

  Object.defineProperty(window, 'getSelection', {
    configurable: true,
    value: () => {
      return new RangeObj()
    }
  })
}

export const loadImage = async (imagePath) =>
  await fs.readFile(path.resolve(imagePath))

export const toHtml = (Component) =>
  ReactDOMServer.renderToStaticMarkup(Component)

export const axeComponent = async (...components) => {
  const html = components
    .filter((Component) =>
      // enzyme names the mounted wrapper: ReactWrapper
      /react/i.test(String(Component.constructor))
    )
    .map((Component) => toHtml(Component))
    .join('\n')

  return await axe(
    `<main>${html}</main>`,
    typeof components[1] === 'object' ? components[1] : null
  )
}

export function attachToBody() {
  let container = document.createElement('div')
  document.body.append(container)
  return container
}

function jestSetup() {}

export default jestSetup
