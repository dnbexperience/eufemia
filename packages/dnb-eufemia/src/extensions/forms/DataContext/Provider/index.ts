export { default } from './Provider'
export * from './Provider'
