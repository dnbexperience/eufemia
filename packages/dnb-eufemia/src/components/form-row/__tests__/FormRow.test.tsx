/**
 * FormRow Test
 *
 */

import React from 'react'
import { axeComponent, loadScss } from '../../../core/jest/jestSetup'
import FormRow, { FormRowProps } from '../FormRow'
import Input from '../../input/Input'
import NumberFormat from '../../number-format/NumberFormat'
import Provider from '../../../shared/Provider'
import { render } from '@testing-library/react'

const props: FormRowProps = {
  label: 'label',
}

describe('FormRow component', () => {
  it('should have vertical direction class', () => {
    render(<FormRow {...props} direction="vertical" />)
    expect(document.querySelector('.dnb-form-row').classList).toContain(
      'dnb-form-row--vertical'
    )
  })

  it('should have an isolated state on nested FormRows', () => {
    render(
      <FormRow vertical>
        <Input label="Vertical" />
        <FormRow vertical={false} label_direction="horizontal">
          <Input label="Horizontal" />
        </FormRow>
        <Input label="Vertical" />
      </FormRow>
    )
    expect(
      document.querySelectorAll('span.dnb-input')[0].classList
    ).toContain('dnb-input--vertical')
    expect(
      document.querySelectorAll('span.dnb-input')[1].classList
    ).toContain('dnb-input--horizontal')
    expect(
      document.querySelectorAll('span.dnb-input')[2].classList
    ).toContain('dnb-input--vertical')
  })

  it('should using formset and legend by default', () => {
    render(<FormRow {...props} />)

    expect(document.querySelector('fieldset')).toBeInTheDocument()
    expect(document.querySelector('legend')).toBeInTheDocument()
  })

  it('should not using formset and legend when no_fieldset is given', () => {
    render(<FormRow {...props} no_fieldset />)
    expect(document.querySelector('label')).toBeInTheDocument()
    expect(document.querySelector('fieldset')).not.toBeInTheDocument()
    expect(document.querySelector('legend')).not.toBeInTheDocument()
  })

  it('should support locale context forwarding', () => {
    const { rerender } = render(
      <FormRow>
        <NumberFormat currency>1234</NumberFormat>
      </FormRow>
    )

    expect(
      document
        .querySelector('.dnb-number-format')
        .querySelector('.dnb-number-format__visible').textContent
    ).toBe('1 234,00 kr')

    rerender(
      <FormRow locale="en-GB">
        <NumberFormat currency>1234</NumberFormat>
      </FormRow>
    )

    expect(
      document
        .querySelector('.dnb-number-format')
        .querySelector('.dnb-number-format__visible').textContent
    ).toBe('NOK 1 234.00')
  })

  it('should not overwrite locale from provider when not set', () => {
    render(
      <Provider locale="en-GB">
        <FormRow>
          <NumberFormat currency>1234</NumberFormat>
        </FormRow>
      </Provider>
    )

    expect(
      document
        .querySelector('.dnb-number-format')
        .querySelector('.dnb-number-format__visible').textContent
    ).toBe('NOK 1 234.00')
  })

  it('should react correct on two states in row', () => {
    const { rerender } = render(
      <FormRow {...props} disabled={false}>
        <Input />
      </FormRow>
    )
    rerender(
      <FormRow {...props} disabled={true}>
        <Input />
      </FormRow>
    )
    expect(document.querySelector('input[disabled]')).toBeInTheDocument()
  })

  it('should validate with ARIA rules', async () => {
    const Comp = render(<FormRow {...props} />)
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('FormRow scss', () => {
  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })

  it('have to match default theme snapshot', () => {
    const css = loadScss(
      require.resolve('../style/themes/dnb-form-row-theme-ui.scss')
    )
    expect(css).toMatchSnapshot()
  })
})
