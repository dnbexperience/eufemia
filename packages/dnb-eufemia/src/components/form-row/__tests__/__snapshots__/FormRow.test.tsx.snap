// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRow scss has to match style dependencies css 1`] = `
"/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
 * FormRow component
 *
 */
/*
 * Utilities
 */
.dnb-form-row {
  align-items: baseline;
}
.dnb-form-row, .dnb-form-row.dnb-section {
  display: flex;
}
.dnb-form-row__content {
  display: inline-flex;
  flex: 1;
  width: 100%;
}
.dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content {
  align-items: baseline;
}
.dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content .dnb-form-label {
  white-space: nowrap;
}
.dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content--wrap {
  display: block;
}
@media screen and (max-width: 40em) {
  .dnb-form-row:not(.dnb-form-row--vertical) {
    flex-direction: column;
    align-items: flex-start;
  }
  .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-label {
    margin-bottom: 0.5rem;
  }
  .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content {
    flex-direction: column;
  }
  .dnb-form-row:not(.dnb-form-row--vertical) > .dnb-form-row__content > .dnb-form-label {
    margin-top: 1rem;
  }
}
@media screen and (min-width: 40.0625em) {
  .dnb-form-row--centered.dnb-form-row:not(.dnb-form-row--vertical) .dnb-form-row__content {
    align-items: center;
  }
}
.dnb-form-row--vertical > .dnb-form-row__label {
  margin-bottom: 0.5rem;
}
.dnb-form-row--vertical > .dnb-form-row__content, .dnb-form-row--vertical.dnb-form-row:not(.dnb-form-row--nested) {
  flex-direction: column;
}
.dnb-form-row--vertical-label {
  flex-direction: column;
}
.dnb-form-row > .dnb-form-status {
  margin: 0.5rem 0;
}
.dnb-form-row > .dnb-form-row__label--vertical {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}
.dnb-form-row > .dnb-form-row__label--vertical ~ .dnb-form-row__content {
  grid-column: column1-start/column1-end;
}
.dnb-form-row__label-dummy {
  margin: 0;
}
.dnb-form-row__fieldset, .dnb-core-style .dnb-form-row__fieldset {
  margin: 0;
  padding: 0;
  border: none;
  width: 100%;
}"
`;

exports[`FormRow scss have to match default theme snapshot 1`] = `
"/*
 * FormRow theme
 *
 */"
`;
